crates/driver/src/domain/competition/mod.rs

```diff
@@ -39,79 +40,130 @@ pub struct Competition {
     pub simulator: Simulator,
     pub now: time::Now,
     pub mempools: Vec<Mempool>,
-    pub settlement: Mutex<Option<(solution::Id, settlement::Verified)>>,
+    pub settlement: Mutex<Option<settlement::Verified>>,
 }
 
 impl Competition {
     /// Solve an auction as part of this competition.
-    pub async fn solve(&self, auction: &Auction) -> Result<(solution::Id, solution::Score), Error> {
+    pub async fn solve(
+        &self,
+        auction: &Auction,
+    ) -> Result<(settlement::Id, solution::Score), Error> {
+        // Fetch the liquidity needed by the auction.
         tracing::trace!("fetching liquidity");
-        let liquidity = self.liquidity.fetch(&Self::liquidity_pairs(auction)).await;
+        let liquidity = self
+            .liquidity
+            .fetch(
+                &auction
+                    .orders
+                    .iter()
+                    .filter_map(|order| match order.kind {
+                        order::Kind::Market | order::Kind::Limit { .. } => {
+                            liquidity::TokenPair::new(order.sell.token, order.buy.token)
+                        }
+                        order::Kind::Liquidity => None,
+                    })
+                    .collect(),
+            )
+            .await;
+
+        // Fetch the solutions from the solver.
         tracing::trace!("solving");
-        let solution = self
+        let solutions = self
             .solver
             .solve(auction, &liquidity, auction.deadline.timeout(self.now)?)
             .await?;
 
-        if solution.is_empty() {
-            // Don't waste resources on simulating an empty solution.
-            return Ok((solution.id, solution::Score::zero()));
-        }
+        // Empty solutions aren't useful, so discard them.
+        let solutions = solutions.into_iter().filter(|solution| {
+            if solution.is_empty() {
+                tracing::info!(id = ?solution.id, "discarding solution: empty");
+                false
+            } else {
+                true
+            }
+        });
+
+        // Verify the solutions, encoding them into settlements.
+        let settlements = join_all(solutions.map(|solution| async move {
+            tracing::trace!(id = ?solution.id, "verifying");
+            (
+                solution.id,
+                solution.verify(&self.eth, &self.simulator, auction).await,
```

Comment:
Not directly related to this PR. I notice now that this can be a problem with the auction deadline. We forward the deadline to the solver. When we get the result, we spend extra time verifying the solution(s) and scoring them. This might push us out of the deadline. This is exasperated with more solutions. Not sure what the best way to handle it is.

Reply:
The deadline is reduced before being sent to the solver for exactly this reason. Perhaps the best option would be to make it so that the amount to reduce by can be specified in the config, and additionally, we should log how long the verification is taking. This way the solvers can see how long verification is taking, and since they know the average number of solution they'll return, they can set the config value accordingly.

Reply:
Making this configurable would be a nice-to-have.

Reply:
https://github.com/cowprotocol/services/issues/1474